Implement lazy member loading for mirrored protocol members in imported types

slavapestov opened the pull request.

slavapestov commented:
@jrose-apple The second patch doesn't have the expected performance gain because while we spend less time in loadAllMembers(), the lookupDirect() path has slowed down because a lazy member 'miss' does so much extra work. So I need to optimize something there. But comments on the general approach are welcome. I might land the first commit separately if all else fails, although it would be a shame.

jrose-apple commented:
This generally makes sense to me. I don't see that that much of the work is avoidable the way things are set up now, so maybe the real answer is going to have to be altering the lookup tables (which is not a bad idea anyway).

slavapestov commented:
@swift-ci Please test compiler performance

jrose-apple commented:
Wow, we dropped **a third** of the Clang decls we imported. I can actually believe that. And yet we also regressed compile time by over 1%.

slavapestov committed:
ClangImporter: Lazily load mirrored protocol members in classes

slavapestov committed:
ClangImporter: Don't force loading of all superclass members in loadNamedMembers()

slavapestov committed:
ClangImporter: Refactor importInheritedConstructors()

The lambda here was completely unnecessary.

CodaFi commented:
Can you strike the inherited constructors part of this patch?  It's a bit noisy.

CodaFi commented on lib/AST/NameLookup.cpp:
```diff
@@ -1143,37 +1143,30 @@ void ExtensionDecl::addedMember(Decl *member) {
                                           MemberLookupTable &LookupTable,
                                           DeclBaseName name,
                                           IterableDeclContext *IDC) {
+  assert(!IDC->isLoadingLazyMembers());
```
Let's see if your assert can stand up to the bots.  You might be in the same position I am right now with the lookup requests.

slavapestov commented:
@CodaFi The inherited constructors change is in a separate NFC commit.

slavapestov committed:
ClangImporter: Fix up importSubscript() to not depend on import order

Once lazy loading supports mirrored protocol members, we can end up
calling importSubscript() for a setter method before we've mirrored
the getter. In this case, we call findCounterpart() with the setter,
which finds the imported getter from the ProtocolDecl itself.

Stop processing the potential subscript in this case, since when we
later mirror the protocol member, we'll go and build it again.

This should be NFC without the next change I'm working on.

slavapestov committed:
ClangImporter: Refactor loadObjCMethods() to not call CollectMultipleMethodsInGlobalPool()

Let's try calling ObjCInterface::lookupMethod() instead. This also
eliminates an order dependency between selector conflict checking
and protocol member mirroring, which fixes a diagnostics regression
once lazy loading is enabled for mirrored protocol members.

slavapestov committed:
Sema: Trigger lazy loading when recording potential selector conflicts

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test

CodaFi commented:
@swift-ci please smoke test compiler performance

CodaFi commented:
@swift-ci please smoke test compiler performance
